File: service/leitura_service.py
from collections import deque
from repository.leitura_repository import salvar_leitura


class processar_leitura:

    def __init__(self, leituras_buffer =5):                             # Buffer para armazenar as últimas leituras (máximo 5), ao definirmos o tamanho, 
        self.buffer = deque(maxlen=leituras_buffer)                     # hora que encher, ele vai começar a remover o mais antigo para adicionar o novo
        self.campos_obrigatorios= ["SOJA", "FEIJAO", "MILHO", "TRIGO"]
                                                                        


    def executar(self, dados):
        print("\n📥 NOVA LEITURA RECEBIDA") # INDICA LEITURA RECEBIDA
        print("Dados:", dados) 

        for campo in self.campos_obrigatorios:
            if campo not in dados:
                raise ValueError(f"Campo '{campo}' é obrigatório")
            
        # converte valores para float e validar
        try: 
            # o k pega cada elemento de self.campos_obrigatorios a CHAVE
            # o dados[k] pega cada valor usando a chave k, ou seja, pega o valor correspondente a cada campo obrigatório
            # o float() converte esse valor para número decimal
            # e o resultado é um dicionário onde as chaves são os nomes dos campos e os valores são os números convertidos
            valores = {k: float(dados[k]) for k in self.campos_obrigatorios}

        except ValueError as e:
            raise ValueError("Todos os valores devem ser numéricos")
        
        self.validar_leitura(valores) # VALIDA OS VALORES DE CADA CULTURA, SE ESTIVER FORA DO LIMITE LEVANTA EXCEÇÃO
        
        #Adiciona no buffer
        self.buffer.append((valores["SOJA"], valores["FEIJAO"], valores["MILHO"], valores["TRIGO"])) 
        print(f"Buffer atual: {len(self.buffer)}")
        
        # só salva quando tiver 5
        if(len(self.buffer) == 5):
            self.calcular_media_movel()
            
            
    def validar_leitura(self, valores):
        soja = valores["SOJA"]
        feijao = valores["FEIJAO"]
        milho = valores["MILHO"]
        trigo = valores["TRIGO"]

        if not (20.0 <= soja <= 35.0):
            raise ValueError("Soja fora do limite (20 - 35)")

        if not (25.0 <= feijao <= 40.0):
            raise ValueError("Feijão fora do limite (25 - 40)")

        if not (15.0 <= milho <= 30.0):
            raise ValueError("Milho fora do limite (15 - 30)")

        if not (30.0 <= trigo <= 50.0):
            raise ValueError("Trigo fora do limite (30 - 50)")


    def calcular_media_movel(self):

        soma_soja = sum(l[0] for l in self.buffer) # pega o primeiro elemento de cada tupla (soja) e soma
        soma_feijao = sum(l[1] for l in self.buffer) # pega o segundo elemento de cada tupla (feijao) e soma
        soma_milho = sum(l[2] for l in self.buffer) # pega o terceiro elemento de cada tupla (milho) e soma
        soma_trigo = sum(l[3] for l in self.buffer) # pega o quarto elemento de cada tupla (trigo) e soma

        media_soja = soma_soja / 5
        media_feijao = soma_feijao / 5
        media_milho = soma_milho / 5
        media_trigo = soma_trigo / 5

        print("\n==============================")
        print("🔥 SALVANDO MÉDIA NO BANCO 🔥")
        print("Soja:", media_soja)
        print("Feijao:", media_feijao)
        print("Milho:", media_milho)
        print("Trigo:", media_trigo)
        print("==============================\n")

        #salva somente a média
        salvar_leitura(media_soja, media_feijao, media_milho, media_trigo)
        

    

        

File: src/service/leitura_service.ts
import { leituraRepository } from '../repository/leitura_repository.js';
import type { Leitura } from '../types/graos/processarLeitura_interface.js';
import {Cultura} from '../types/graos/processarLeitura_interface.js'


export class ProcessarLeitura {

    private repo = new leituraRepository()
    private buffer: Leitura[] = [];
    private readonly maxLen: number;
    
    // Usamos o Enum para definir os campos obrigatórios
    private readonly camposObrigatorios: Cultura[] = [
        Cultura.SOJA, 
        Cultura.FEIJAO, 
        Cultura.MILHO, 
        Cultura.TRIGO
    ];

    constructor(leiturasBuffer: number = 5) {
        this.maxLen = leiturasBuffer;
    }

    public async executar(dados: any): Promise<void> {
        console.log("\n📥 NOVA LEITURA RECEBIDA");
        
        // 1. Validação e Conversão (Garantindo que os dados sigam a interface Leitura)
        const valores = this.validarEConverter(dados);

        // 2. Validação de Regras de Negócio (Limites)
        this.validarLimites(valores);

        // 3. Adiciona no Buffer (Lógica de deque)
        this.buffer.push(valores);
        if (this.buffer.length > this.maxLen) {
            this.buffer.shift();
        }

        console.log(`Buffer atual: ${this.buffer.length}/${this.maxLen}`);

        // 4. Salva a média quando o buffer encher
        if (this.buffer.length === this.maxLen) {
            await this.calcularMediaMovel();
        }
    }

    private validarEConverter(dados: any): Leitura {
        const resultado: Partial<Leitura> = {};

        for (const campo of this.camposObrigatorios) {
            if (!(campo in dados)) {
                throw new Error(`Campo '${campo}' é obrigatório`);
            }

            const valorNumerico = Number(dados[campo]);
            if (isNaN(valorNumerico)) {
                throw new Error(`Valor do campo '${campo}' deve ser numérico`);
            }
            // TypeScript entende que 'campo' é uma chave válida de Leitura
            resultado[campo] = valorNumerico;
        }

        return resultado as Leitura;
    }

    private validarLimites(v: Leitura): void { 
        // Usando sua interface Leitura com segurança de tipos
        if (v.SOJA < 20 || v.SOJA > 35) throw new Error("Soja fora do limite (20 - 35)");
        if (v.FEIJAO < 25 || v.FEIJAO > 40) throw new Error("Feijão fora do limite (25 - 40)");
        if (v.MILHO < 15 || v.MILHO > 30) throw new Error("Milho fora do limite (15 - 30)");
        if (v.TRIGO < 30 || v.TRIGO > 50) throw new Error("Trigo fora do limite (30 - 50)");
    }

    private async calcularMediaMovel(): Promise<void> {
        const total = this.buffer.reduce((acc, curr) => ({
            SOJA: acc.SOJA + curr.SOJA,
            FEIJAO: acc.FEIJAO + curr.FEIJAO,
            MILHO: acc.MILHO + curr.MILHO,
            TRIGO: acc.TRIGO + curr.TRIGO
        }), { SOJA: 0, FEIJAO: 0, MILHO: 0, TRIGO: 0 } as Leitura);

        const mediaSoja = total.SOJA / this.maxLen;
        const mediaFeijao = total.FEIJAO / this.maxLen;
        const mediaMilho = total.MILHO / this.maxLen;
        const mediaTrigo = total.TRIGO / this.maxLen;

        console.log("\n SALVANDO MÉDIA NO BANCO ");
        console.log(`Média Soja: ${mediaSoja.toFixed(2)}`);
        console.log(`Média Feijão: ${mediaFeijao.toFixed(2)}`);
        console.log(`Média Milho: ${mediaMilho.toFixed(2)}`);
        console.log(`Média Trigo: ${mediaTrigo.toFixed(2)}`);
        
        // Mantendo a chamada original do seu repository
        await this.repo.salvarLeitura(mediaSoja, mediaFeijao, mediaMilho, mediaTrigo);
    }
}